src/group_theory/nilpotent.lean

```diff
@@ -0,0 +1,284 @@
+* `lower_central_series G : ℕ → subgroup G` : the lower central series of a group `G`.
+     This is a decreasing sequence of subgroups `H n` of `G` starting at `⊤` and
+     with the property that `H (n + 1) = ⁅H n, G⁆`.
+* `is_nilpotent` : A group G is nilpotent if its upper central series reaches `⊤`, or
+    equivalently if its lower central series reaches `⊥`.
+* `nilpotency_class` : the length of the upper central series of a nilpotent group.
+* `is_ascending_central_series (H : ℕ → subgroup G) : Prop` and
+* `is_descending_central_series (H : ℕ → subgroup G) : Prop` : Note that in the literature
+  a "central series" for a group is usually defined to be a *finite* sequence of normal subgroups
+  `H 0`, `H 1`, ..., starting at `⊤`, finishing at `⊥`, and with each `H n / H (n + 1)`
+  central in `G / H (n + 1)`. In this formalisation it is convenient to have two weaker predicates
+  on an infinite sequence of subgroups `H n` of `G`: we say this sequence is a descending central
+  series if it starts at `G` and `⁅H n, ⊤⁆ ⊆ H (n + 1)` for all `n`. Note that this series
+  may not terminate at `⊥`, and the `H i` need not be normal. Similarly a sequence is an ascending
+  central series if `H 0 = ⊥` and `⁅H (n + 1), ⊤⁆ ⊆ H n` for all `n`, with no requirement
+  that the series reaches `⊤` or that the `H i` are normal.
+
+## Main theorems
+
+Recall that `G` is *defined* to be nilpotent if the upper central series reaches `⊤`.
+* `nilpotent_iff_finite_ascending_central_series` : `G` is nilpotent iff some ascending central
+    series reaches `⊤`.
+* `nilpotent_iff_finite_descending_central_series` : `G` is nilpotent iff some descending central
+    series reaches `⊥`.
+* `nilpotent_iff_lower` : `G` is nilpotent iff the lower central series reaches `⊥`.
+
+## Warning
+
+A "central series" is usually defined to be a finite sequence of normal subgroups going
+from `⊥` to `⊤` with the property that each subquotient is contained within the centre of
+the associated quotient of `G`. This means that if `G` is not nilpotent, then
+none of what we have called `upper_central_series G`, `lower_central_series G` or
+the sequences satisfying `is_ascending_central_series` or `is_descending_central_series`
+are actually central series.
+
+-/
+variables (G : Type*) [group G]
+
+open subgroup
+
+/-- `upper_central_series G n` is the `n`th term in the upper central series of `G`. -/
+def upper_central_series : ℕ → subgroup G
+| 0 := ⊥
+| (n+1) := subgroup.normal_closure {x : G | ∀ y : G, x * y * x⁻¹ * y⁻¹ ∈ upper_central_series n}
+
+instance (G : Type*) [group G] : ∀ (n : ℕ), normal (upper_central_series G n)
+| 0 := subgroup.bot_normal
+| (n+1) := subgroup.normal_closure_normal
+
+lemma upper_central_series_zero_def (G : Type*) [group G] : upper_central_series G 0 = ⊥ := rfl
+
+/-- An auxiliary definition, which ultimately will equal the upper central series. -/
+private def upper_central_series_aux : ℕ → subgroup G
```

Comment:
Is there a reason you don't use this definition directly?

Reply:
I imagine you could avoid needing both with something like
```suggestion
private def upper_central_series_aux : ℕ → Σ' H : subgroup G, normal H
```
So that you have the proofs of normalness as part of your recursion

Reply:
As you probably realise, what's going on here is that the books say "here is the recursive definition of this sequence of normal subgroups, we will define the subgroups as subsets and then prove they're normal subgroups by induction all at the same time". The trick I used to get around this was to make sure they're normal subgroups from the outset by saying "it's the normal subgroup generated by this set" and then later on showing that the set itself was a normal subgroup -- note that the definition of upper_central_series_aux depends on upper_central_series. My decision was based on nothing more than readability. My definition of upper central series is short, clear, and any mathematician who knows the theory will see that it's trivially equivalent to the usual definition (the aux one). The definition of the aux one is intimidating to the group theorist who knows no Lean. This was my only reason. I am happy to refactor if you think that removal of private aux definitions is more important than readabiity of a definition (I mean, our definition of group is now completely unreadable and this doesn't bother me, it's just changed the way I teach it).

Reply:
In this case my main motivation was to make `mem_upper_central_series_succ_iff` true by definition. I think you could still make this read like the books by having the definition be
```lean
-- the recursive step
private def upper_central_series_step (H : subgroup G) [H.normal] : subgroup G :=
{ carrier := {x : G | ∀ y : G, x * y * x⁻¹ * y⁻¹ ∈ H, ..sorry}

instance (H : subgroup G) [H.normal] : : (upper_central_series_step H).normal := sorry

private def upper_central_series_aux : ℕ → Σ' H : subgroup G, normal H
| 0 := ⟨⊥, sorry⟩
| (n + 1) :=
   let un := upper_central_series_aux n,  un_g = un.1, un_normal := un.2 in
   by exactI ⟨upper_central_series_step un, by apply_instance⟩
```

Reply:
I did this (I am unclear about how much definitional equality will be important but my instinct is that it's probably a good idea)